onnx/defs/tensor/defs.cc

```diff
@@ -1794,4 +1794,132 @@ ONNX_OPERATOR_SET_SCHEMA(
+            AttributeProto::INT,
+            OPTIONAL)
+        .Input(0, "X", "A N-D input tensor that is to be processed.", "T")
+        .Output(0, "Y", "A 1-D tensor of the same type as 'X' " 
+                        "containing all the unique values in 'X', either sorted " 
+                        "or maintained in the same order they occur in the input 'X'", "T")
+        .Output(1, "indices", "A 1-D INT64 tensor " 
+                          "containing indices of 'X' for each unique element in 'Y'",
+                          "tensor(int64)")
+        .Output(2, "inverse_indices", "A 1-D INT64 tensor "
+                          "containing indices of 'Y' for each input element in 'X'",
+                          "tensor(int64)")
+        .Output(3, "counts", "A 1-D INT64 tensor containing the "
+                             "the count of each element "
+                             "of 'Y' in the input 'X'", "tensor(int64)")
+        .TypeConstraint("T", OpSchema::all_tensor_types(), "Input can be of any tensor type.")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) { 
+          // Type inference
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          const TypeProto* xTensorProto = ctx.getInputType(0);
+          TypeProto* yTensorProto = ctx.getOutputType(0);
+          TypeProto* indicesTensorProto = nullptr;
+          TypeProto* inverseIndicesTensorProto = nullptr;
+          TypeProto* countsTensorProto = nullptr;
+          
+          // 'indices', 'inverse_indices', and 'counts' are 1-D tensors of unknown dimension.
+          auto num_outputs = ctx.getNumOutputs();
+          if (num_outputs >= 2) {
+            indicesTensorProto = ctx.getOutputType(1);
+            updateOutputElemType(ctx, 1, TensorProto::INT64);
+            indicesTensorProto->mutable_tensor_type()->mutable_shape()->add_dim();
+          }
+
+          if (num_outputs >= 3) {
+            inverseIndicesTensorProto = ctx.getOutputType(2);
+            updateOutputElemType(ctx, 2, TensorProto::INT64);
+            inverseIndicesTensorProto->mutable_tensor_type()->mutable_shape()->add_dim();
+          }
+
+          if (num_outputs >= 4) {
+            countsTensorProto = ctx.getOutputType(3);
+            updateOutputElemType(ctx, 3, TensorProto::INT64);
+            countsTensorProto->mutable_tensor_type()->mutable_shape()->add_dim();
+          }
+
+          auto axisAttr = ctx.getAttribute("axis");
+          if (!axisAttr) {
+            // 'axis' is not provided. Input 'X' is flattened.
+            // 'Y' is a 1-D tensor of unknown dimension.
+            yTensorProto->mutable_tensor_type()->mutable_shape()->add_dim();
+          }
+          else 
+          {
+            // 'axis' is provided.
+            int axis = static_cast<int>(axisAttr->i());
+
+            // 'Y' has the same shape as 'X' except in the 'axis' dimension which is unknown.
```

Comment:
When I read the scipy documentation, it says that all the other dimensions (other than axis) are flattened ... is that correct, or does that flattening not happen for the output?

Reply:
I think it means that the rest dimensions are flattened after the 'axis' dimension is moved to the front for sorting. However the final output is transposed back to the original input form. Test cases also conform this.